src/file_drop/mod.rs

```diff
@@ -0,0 +1,147 @@
+#[cfg(target_os = "linux")]
+mod linux;
+#[cfg(target_os = "linux")]
+pub(crate) use linux::*;
+#[cfg(target_os = "macos")]
+mod macos;
+#[cfg(target_os = "macos")]
+pub(crate) use macos::*;
+#[cfg(target_os = "windows")]
+mod win;
+#[cfg(target_os = "windows")]
+pub(crate) use win::*;
+
+use core::fmt;
+use std::{cell::Cell, path::PathBuf, sync::Arc};
+
+#[derive(Debug, Serialize, Clone)]
+/// The status of a file drop event.
+pub enum FileDropStatus {
+    /// The file(s) have been dragged onto the window, but have not been dropped yet.
+    Hovered(Vec<PathBuf>),
+
+    /// The file(s) have been dropped onto the window.
+    Dropped(Vec<PathBuf>),
+
+    /// The file(s) drop was aborted.
+    Cancelled(Vec<PathBuf>),
+}
+
+// This needs to be defined because internally the respective events do not always yield a PathBuf.
+// We can easily remember what was cancelled though, as Hovered and Dropped events will always yield a PathBuf which we will save ourselves for later reference.
+pub(crate) enum FileDropEvent {
+    Hovered,
+    Dropped,
+    Cancelled
+}
+
+#[derive(Clone)]
+pub struct FileDropHandler {
```

Comment:
Are you sure? I did this so that we could still derive Debug.

Reply:
I usually just use debuggers. If users want debug trait, they can always define new type and implement it

Reply:
Debugger support is crap on Windows for Rust unfortunately. They usually do not capture the data properly at all. I think this is a user-friendly way, too.

However... I've just remembered why I actually did this, it's because of `Attributes` which derives Debug and Clone. Can't clone a Fn without Box-ing it which I feel isn't very elegant on the user's pov. It's this:

```rust
Attributes {
    ...
    file_drop_handler: Arc::new(Box::new(|status| { ... })),
    ...
}
```

vs this:
```rust
Attributes {
    ...
    file_drop_handler: FileDropHandler::new(|status| { ... }),
    ...
}
```

Reply:
I think no matter what Arc<Box<...>> is required here. The platforms differ with how they handle file dropping and I personally experienced multithreading hell on macOS. I think using an Arc would be safest.

Reply:
Let's stay as is then. I could look deeper later and see if anything can be improved.